src/decoder/lattice-faster-online-decoder.cc

```diff
@@ -135,24 +135,28 @@ typename LatticeFasterOnlineDecoderTpl<FST>::BestPathIterator LatticeFasterOnlin
   Token *tok = static_cast<Token*>(iter.tok);
   int32 cur_t = iter.frame, ret_t = cur_t;
   if (tok->backpointer != NULL) {
+    BaseFloat best_graph_cost = std::numeric_limits<BaseFloat>::infinity();
     ForwardLinkT *link;
     for (link = tok->backpointer->links;
          link != NULL; link = link->next) {
       if (link->next_tok == tok) { // this is the link to "tok"
-        oarc->ilabel = link->ilabel;
-        oarc->olabel = link->olabel;
-        BaseFloat graph_cost = link->graph_cost,
-            acoustic_cost = link->acoustic_cost;
-        if (link->ilabel != 0) {
-          KALDI_ASSERT(static_cast<size_t>(cur_t) < this->cost_offsets_.size());
-          acoustic_cost -= this->cost_offsets_[cur_t];
-          ret_t--;
+        BaseFloat graph_cost = link->graph_cost;
+        if (graph_cost < best_graph_cost) {
+          oarc->ilabel = link->ilabel;
+          oarc->olabel = link->olabel;
+          BaseFloat acoustic_cost = link->acoustic_cost;
+          if (link->ilabel != 0) {
+            KALDI_ASSERT(static_cast<size_t>(cur_t) < this->cost_offsets_.size());
+            acoustic_cost -= this->cost_offsets_[cur_t];
+            ret_t--;
+          }
+          oarc->weight = LatticeWeight(graph_cost, acoustic_cost);
+          best_graph_cost = graph_cost;
         }
-        oarc->weight = LatticeWeight(graph_cost, acoustic_cost);
-        break;
       }
     }
-    if (link == NULL) { // Did not find correct link.
+    if (link == NULL &&
```

Comment:
The post condition of the above "for loop termination" ensures (link == NULL), so here the first condition seems to be redundent.